File: packages/lynx-updates/ios/UpdateDownloader.swift
import Foundation
import CryptoKit

/// Streaming bundle downloader: bytes go straight to `tmp/<id>.partial`
/// with an incremental SHA-256, then atomically move into `updates/<id>/`
/// once the hash matches. Single-flight — concurrent calls beyond the first
/// fail fast.
final class UpdateDownloader: NSObject, URLSessionDataDelegate {

    private static let inFlightLock = NSLock()
    private static var inFlight = false

    private let partialURL: URL
    private var output: FileHandle?
    private var hasher = SHA256()
    private var receivedBytes: Int64 = 0
    private var totalBytes: Int64?
    private var lastProgressAt = Date.distantPast
    private var result: String? = "Download did not complete"
    private let done = DispatchSemaphore(value: 0)

    private init(partialURL: URL) {
        self.partialURL = partialURL
    }

    /// Synchronous (call off the main thread). Returns nil on success or an
    /// error message (prefixed with E_* codes the module maps to the bridge).
    static func download(
        url: String,
        expectedSha256: String,
        updateId: String,
        headers: [String: String],
        manifestJson: String,
    ) -> String? {
        let store = UpdateStore.shared

        // Already on disk and intact → success without a byte transferred.
        if FileManager.default.fileExists(atPath: store.bundleFile(updateId).path),
           store.verifySha256(updateId) {
            return nil
        }

        inFlightLock.lock()
        if inFlight {
            inFlightLock.unlock()
            return "E_DOWNLOAD_IN_PROGRESS: another download is running"
        }
        inFlight = true
        inFlightLock.unlock()
        defer {
            inFlightLock.lock()
            inFlight = false
            inFlightLock.unlock()
        }

        guard let requestURL = URL(string: url) else {
            return "Download failed: invalid URL \(url)"
        }

        let fm = FileManager.default
        try? fm.createDirectory(at: store.tmpDir, withIntermediateDirectories: true)
        let partial = store.tmpDir.appendingPathComponent("\(updateId).partial")
        fm.createFile(atPath: partial.path, contents: nil)

        let downloader = UpdateDownloader(partialURL: partial)
        guard let handle = try? FileHandle(forWritingTo: partial) else {
            return "Download failed: cannot open staging file"
        }
        downloader.output = handle

        var request = URLRequest(url: requestURL, timeoutInterval: 30)
        for (key, value) in headers {
            request.setValue(value, forHTTPHeaderField: key)
        }
        let session = URLSession(configuration: .ephemeral, delegate: downloader, delegateQueue: nil)
        session.dataTask(with: request).resume()
        downloader.done.wait()
        session.finishTasksAndInvalidate()

        if let failure = downloader.result {
            try? fm.removeItem(at: partial)
            return failure
        }

        let actual = downloader.hasher.finalize().map { String(format: "%02x", $0) }.joined()
        guard actual == expectedSha256.lowercased() else {
            try? fm.removeItem(at: partial)
            return "E_HASH_MISMATCH: expected \(expectedSha256), got \(actual)"
        }

        // Promote: metadata first, bundle move last.
        let dir = store.updateDir(updateId)
        try? fm.removeItem(at: dir)
        do {
            try fm.createDirectory(at: dir, withIntermediateDirectories: true)
            var meta = (try? JSONSerialization.jsonObject(with: Data(manifestJson.utf8))) as? [String: Any] ?? [:]
            meta["sha256"] = expectedSha256.lowercased()
            meta["sizeBytes"] = downloader.receivedBytes
            meta["sourceUrl"] = url
            meta["downloadedAt"] = Int(Date().timeIntervalSince1970 * 1000)
            let metaData = try JSONSerialization.data(withJSONObject: meta)
            try metaData.write(to: store.updateJsonFile(updateId), options: .atomic)
            try fm.moveItem(at: partial, to: store.bundleFile(updateId))
        } catch {
            try? fm.removeItem(at: partial)
            return "Download failed: \(error.localizedDescription)"
        }
        return nil
    }

    // MARK: - URLSessionDataDelegate

    func urlSession(
        _ session: URLSession, dataTask: URLSessionDataTask,
        didReceive response: URLResponse,
        completionHandler: @escaping (URLSession.ResponseDisposition) -> Void,
    ) {
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            result = "Download failed: HTTP \(http.statusCode)"
            completionHandler(.cancel)
            return
        }
        totalBytes = response.expectedContentLength >= 0 ? response.expectedContentLength : nil
        completionHandler(.allow)
    }

    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
        output?.write(data)
        hasher.update(data: data)
        receivedBytes += Int64(data.count)
        let now = Date()
        if now.timeIntervalSince(lastProgressAt) >= 0.15 {
            lastProgressAt = now
            UpdatesEventBus.shared.emitProgress(receivedBytes: receivedBytes, totalBytes: totalBytes)
        }
    }

    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        try? output?.close()
        if let error {
            // A cancel from didReceive(response:) already set a specific message.
            if result == "Download did not complete" || result == nil {
                result = "Download failed: \(error.localizedDescription)"
            }
        } else {
            result = nil
            UpdatesEventBus.shared.emitProgress(
                receivedBytes: receivedBytes, totalBytes: totalBytes ?? receivedBytes)
        }
        done.signal()
    }
}

File: packages/lynx-updates/android/com/sigx/updates/UpdateDownloader.kt
package com.sigx.updates

import android.content.Context
import android.util.Log
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.util.concurrent.atomic.AtomicBoolean
import org.json.JSONObject

/**
 * Streaming bundle downloader: bytes go straight to `tmp/<id>.partial` with
 * an incremental SHA-256, then atomically move into `updates/<id>/` once the
 * hash matches. Single-flight — concurrent calls beyond the first fail fast.
 */
object UpdateDownloader {

    private const val TAG = "SigxUpdates"
    private const val CONNECT_TIMEOUT_MS = 15_000
    private const val READ_TIMEOUT_MS = 30_000
    private const val PROGRESS_INTERVAL_MS = 150L

    private val inFlight = AtomicBoolean(false)

    /**
     * @return null on success, or an error message. `code` semantics ride in
     *         the message prefix; the module maps them to bridge errors.
     */
    fun download(
        context: Context,
        url: String,
        expectedSha256: String,
        updateId: String,
        headers: Map<String, String>,
        manifestJson: String,
    ): String? {
        // Already on disk and intact → success without a byte transferred.
        if (UpdateStore.bundleFile(context, updateId).exists() &&
            UpdateStore.verifySha256(context, updateId)
        ) {
            return null
        }

        if (!inFlight.compareAndSet(false, true)) {
            return "E_DOWNLOAD_IN_PROGRESS: another download is running"
        }
        try {
            return downloadLocked(context, url, expectedSha256, updateId, headers, manifestJson)
        } finally {
            inFlight.set(false)
        }
    }

    private fun downloadLocked(
        context: Context,
        url: String,
        expectedSha256: String,
        updateId: String,
        headers: Map<String, String>,
        manifestJson: String,
    ): String? {
        val tmpDir = UpdateStore.tmpDir(context)
        tmpDir.mkdirs()
        val partial = File(tmpDir, "$updateId.partial")

        var connection: HttpURLConnection? = null
        try {
            connection = URL(url).openConnection() as HttpURLConnection
            connection.connectTimeout = CONNECT_TIMEOUT_MS
            connection.readTimeout = READ_TIMEOUT_MS
            connection.instanceFollowRedirects = true
            for ((key, value) in headers) {
                connection.setRequestProperty(key, value)
            }

            val status = connection.responseCode
            if (status !in 200..299) {
                return "Download failed: HTTP $status"
            }

            val totalBytes = connection.contentLengthLong.takeIf { it >= 0 }
            val digest = MessageDigest.getInstance("SHA-256")
            var receivedBytes = 0L
            var lastProgressAt = 0L

            connection.inputStream.use { input ->
                FileOutputStream(partial).use { out ->
                    val buffer = ByteArray(64 * 1024)
                    while (true) {
                        val read = input.read(buffer)
                        if (read < 0) break
                        out.write(buffer, 0, read)
                        digest.update(buffer, 0, read)
                        receivedBytes += read
                        val now = System.currentTimeMillis()
                        if (now - lastProgressAt >= PROGRESS_INTERVAL_MS) {
                            lastProgressAt = now
                            UpdatesEventBus.emitProgress(receivedBytes, totalBytes)
                        }
                    }
                    out.fd.sync()
                }
            }
            UpdatesEventBus.emitProgress(receivedBytes, totalBytes ?: receivedBytes)

            val actual = digest.digest().joinToString("") { "%02x".format(it) }
            if (!actual.equals(expectedSha256, ignoreCase = true)) {
                partial.delete()
                return "E_HASH_MISMATCH: expected $expectedSha256, got $actual"
            }

            // Atomic-ish promote: write metadata first, bundle rename last.
            val dir = UpdateStore.updateDir(context, updateId)
            dir.deleteRecursively()
            dir.mkdirs()
            val meta = try {
                JSONObject(manifestJson)
            } catch (e: Exception) {
                JSONObject()
            }
            meta.put("sha256", expectedSha256.lowercase())
            meta.put("sizeBytes", receivedBytes)
            meta.put("sourceUrl", url)
            meta.put("downloadedAt", System.currentTimeMillis())
            // Atomic + fsynced: a truncated update.json would fail
            // verifySha256() on next launch and roll back a perfectly good
            // bundle as "corrupt".
            val metaFile = UpdateStore.updateJsonFile(context, updateId)
            val metaTmp = File(dir, "update.json.tmp")
            FileOutputStream(metaTmp).use { out ->
                out.write(meta.toString().toByteArray(Charsets.UTF_8))
                out.fd.sync()
            }
            if (!metaTmp.renameTo(metaFile)) {
                metaFile.delete()
                metaTmp.renameTo(metaFile)
            }
            val bundle = UpdateStore.bundleFile(context, updateId)
            if (!partial.renameTo(bundle)) {
                partial.copyTo(bundle, overwrite = true)
                partial.delete()
            }
            Log.i(TAG, "Downloaded update $updateId ($receivedBytes bytes)")
            return null
        } catch (e: Exception) {
            partial.delete()
            return "Download failed: ${e.message}"
        } finally {
            connection?.disconnect()
        }
    }
}
